oak/common/pkcs8.cc

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2021 The Project Oak Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "oak/common/pkcs8.h"
+
+#include <cstring>
+
+#include "absl/strings/escaping.h"
+#include "glog/logging.h"
+
+namespace oak {
+
+namespace pkcs8 {
+
+// Template of the form |prefix|middle| for serializing and deserializing Ed25519 keys as
+// unencrypted PKCS#8 byte arrays. With this template, the 32-byte
+// private key (or seed) will be placed after the `prefix`, and the 32-byte public key will be
+// inserted at the end, resulting in |prefix|private-key|middle|public-key|. With this template,
+// no additional attributes can be included in the encoded byte array. This template should be used
+// in combination with `kEd25519Pkcs8Template`, which specifies the location of the split between
+// `prefix` and `middle`.
+//
+// This is copied from the Rust ring crate
+// https://github.com/briansmith/ring/blob/main/src/ec/curve25519/ed25519/ed25519_pkcs8_v2_template.der.
+const char kBase64Ed25519Pkcs8Template[] = "MFMCAQEwBQYDK2VwBCIEIKEjAyEA";
```

Comment:
I don't really understand why there is a single string which is only used to be split into two other strings, and also why it is base64 encoded rather than just hex-escaped. But if this makes it easier to compare this logic with the one from `ring` I guess it makes sense to keep as it is?

Reply:
Yes. My intention is to keep this implementation similar to the one in `ring`. Using hex-escaped would not be any better, as we still need a conversion, this time using `absl::HexStringToBytes`.

Reply:
I mean actually hex escaped, not hex encoded.

https://stackoverflow.com/questions/45612822/how-to-properly-add-hex-escapes-into-a-string-literal/45612909

Basically the bytes should just be inlined in the resulting binary, there is no need for any runtime conversion.

Reply:
Do you mean like this: 
```
const char kBase64Ed25519Pkcs8Template[] =
    "\x30\x53\x02\x01\x01\x30\x05\x06\x03\x2b\x65\x70\x04\x22\x04\x20\xa1\x23\x03\x21\x00";
```

The embedded null at the end gets stripped. How can that be fixed without going back to byte arrays?

Reply:
Yes that's what I meant. I don't think the null actually gets stripped, it just gets considered as an early string terminator by whatever library it's being passed to that expects C-style strings, but here it should not be treated as such, I guess the length should be provided explicitly somewhere else (but if that's complicated perhaps it's not worth it).

Reply:
Actually, even better: https://stackoverflow.com/a/883683

Reply:
> Actually, even better: https://stackoverflow.com/a/883683

This is the same as the string literal above. The zero at the end is  interpreted as the end of the string when  instantiating `Template`. 

I can initialize `kBase64Ed25519Pkcs8Template` like this:
```
const std::string kBase64Ed25519Pkcs8Template = std::string(
    "\x30\x53\x02\x01\x01\x30\x05\x06\x03\x2b\x65\x70\x04\x22\x04\x20\xa1\x23\x03\x21\x00", 21);
```
(21 is the actual length of the string/byte array)
But still no constant initializer. Should I go for this? Is this actually better than the base64 version?